### PR TITLE
Bluetooth: Host: Don't do `tx_notify` in `conn_recv`

### PR DESCRIPTION
**Pull request by alwa-nordic**
TLDR: The removed code could not guarantee its purpose and was prone to deadlocks.

Digging in git history:

The code, introduced in https://github.com/zephyrproject-rtos/zephyr/pull/21056, was there to make sure the Host has invoked all pending HCI Number of Packets Completed callbacks before accepting new ACL data.

This was done to ensure the higher layers are informed of a successful TX by the Controller before any ACL data that is the result of that TX arrives at the application. Concretely, this was used by ATT to detect pipelined requests from the peer.

ATT pipelining detection was added in https://github.com/zephyrproject-rtos/zephyr/commit/4a57bf6e6cd1e4b1fad87ff2611ff604800ec7b6.

However this relies on the Controller delivering the Number of Packets Completed event promptly, which it is not required to do by spec. So, while this happens to work with some Controllers, there is no guarantee.

When we found some Controller implementations that did not deliver the Number of Packets Completed event promptly, we removed the ATT pipelining detection in https://github.com/zephyrproject-rtos/zephyr/pull/66012. But we forgot to remove the call to `tx_notify` in `conn_recv`, which suffers from the same lack of guarantee.

It used to be that BT RX thread would perform the work of `tx_notify` itself when it needed it to be done. This was changed in https://github.com/zephyrproject-rtos/zephyr/pull/68394. After this, the code forms a dependency from the BT RX thread to the system work queue. Dependencies like this are a liability and should not exist unless necessary. If the system work queue is busy with long-running work, this delays the processing of the ACL data. And if the system work queue is blocked, e.g. by a call into Bluetooth API, this can cause a deadlock.

There is probably more to add to this, but I think this change is sufficiently motivated already. Removing this code should be fine because of the lack of guarantee that it does anything useful.

### CODE REVIEW

**Comment by jhedberg** [comment]
Looks good to me. Any idea why the Babblesim check is failing?

**Comment by alwa-nordic** [comment]
> Looks good to me. Any idea why the Babblesim check is failing?

`tests/bsim/bluetooth/audio/test_scripts/mcs_mcc.sh` is failing. I had a look and did not understand it. I'll dig a bit deeper. @Thalley, do you have any ideas?

**Comment by Thalley** [comment]
> > Looks good to me. Any idea why the Babblesim check is failing?
> 
> `tests/bsim/bluetooth/audio/test_scripts/mcs_mcc.sh` is failing. I had a look and did not understand it. I'll dig a bit deeper. @Thalley, do you have any ideas?


Running the test locally shows that the the `l2cap_sent` callback in OTS is not called. 

The logs from the BSIM from this PR:
```
d_00: @00:00:06.531220  Reading Icon Object succeeded
d_00: @00:00:06.531220  [00:00:06.531,219] <dbg> bt_mcc: bt_mcc_read_icon_url: conn is NULL
d_01: @00:00:06.548827  [00:00:06.548,797] <dbg> bt_ots: l2cap_disconnected: Channel 0x84c34fc disconnected
d_01: @00:00:06.549259  [00:00:06.549,255] <dbg> bt_mcs: read_icon_url: Icon URL read, offset: 0, len:127, URL: http://server.some.where/path/icon.png
d_00: @00:00:06.549518  [00:00:06.549,499] <dbg> bt_ots: l2cap_disconnected: Channel 0x84c7f68 disconnected
```

When running locally without this change:
```
d_00: @00:00:06.531220  Reading Icon Object succeeded
d_00: @00:00:06.531220  [00:00:06.531,219] <dbg> bt_mcc: bt_mcc_read_icon_url: conn is NULL
d_01: @00:00:06.548827  [00:00:06.548,797] <dbg> bt_ots: l2cap_sent: Outgoing data channel 0x82543fc transmitted
d_01: @00:00:06.548827  [00:00:06.548,797] <dbg> bt_ots: l2cap_sent: Scheduled TX on L2CAP CoC is complete
d_01: @00:00:06.548827  [00:00:06.548,797] <dbg> bt_ots: oacp_read_proc_cb: OACP Read Op over L2CAP is completed
d_01: @00:00:06.548827  [00:00:06.548,797] <dbg> bt_mpl: on_object_send: Object Id 0x000000000100, offset 127, length 0
d_01: @00:00:06.548827  [00:00:06.548,797] <dbg> bt_mpl: on_object_send: Read complete
d_01: @00:00:06.548827  [00:00:06.548,797] <dbg> bt_ots: l2cap_disconnected: Channel 0x82543fc disconnected
d_01: @00:00:06.549259  [00:00:06.549,255] <dbg> bt_mcs: read_icon_url: Icon URL read, offset: 0, len:127, URL: http://server.some.where/path/icon.png
d_00: @00:00:06.549518  [00:00:06.549,499] <dbg> bt_ots: l2cap_disconnected: Channel 0x8256e88 disconnected
```

If the `sent` callback not called, then `oacp_read_proc_cb` is not called and the state of the object is not set to `BT_GATT_OTS_OBJECT_IDLE_STATE`.

**Comment by rugeGerritsen** [comment]
We should probably add a test to prove that we have fixed a problem

**Comment by alwa-nordic** [comment]
I think audio is doing something similar in nature to ATT pipelining detection, which as described in the PR description, we had to remove.

The audio code is expecting the `bt_l2cap_chan_ops.sent` event before the `bt_l2cap_chan_ops.disconnected` event. But this is not guaranteed.

The disconnected event is caused by the RX of an L2CAP Disconnection Request control packet in this case.

Now look at the documentation of `bt_l2cap_chan_ops.sent`:

https://github.com/zephyrproject-rtos/zephyr/blob/200de7c00a0a1beb905ec8565a24fce1d498e13b/include/zephyr/bluetooth/l2cap.h#L367-L376

Aside: I think the above documentation is not technically correct. Generically, this event is that the controller produced a Number of Packets Completed event, giving us a buffer credit for a specific connection. For each credit we receive, we invoke one callback from the FIFO of sent packets. It does not mean anything beyond that in general. For example, the controller could give a credit for a connection without having sent anything on that connection yet, maybe because some other process freed some memory. Any meaning assigned to credit events beyond the availability of a buffer is controller specific. We should document this.

But ignoring the aside above, the documentation states that "When the controller does so is implementation dependent". So, expecting this callback before some RX is not guaranteed to work with a generic Controller.

Also, the assertion could be triggered on purpose by the remote, so it's a DOS vector.

I'm still thinking about this problem, but inputs and critique is welcome.

**Comment by alwa-nordic** [comment]
> Alternatively we define the order of events so that all outstanding RX and TX are always finalized before the `disconnected` callback (perhaps for both L2CAP, ISO and ACL), and make sure that all Zephyr-defined implementations follow this order.

You mean make our dialect of HCI with these requirements? That sounds good from a Host perspective. But will this be accepted by the community? I suspect there will be a lot of friction.

> If we do not want to define a specific order of events

If it was clear what the order of events should be, then we can define that, and we'll make the Host so it conserves this order. But it is out of our hands if the Controller reorders these events. And so long as the Controller is allowed to reorder the events, I question the point of the Host trying to conserve it.

If we can define requirements for an ordering with respect to the HCI messages, like the ACL data and Num Complete events, I would really like that. I think we should, because will make our APIs more understandable. It would really help us if application and Host developers negotiated some contracts.

@Thalley, would something the following be sensible?

  - `bt_l2cap_chan_ops.disconnected` comes after `bt_l2cap_chan_ops.sent` had been invoked for each successful return from `bt_l2cap_chan_send()`.
  - `bt_l2cap_chan_ops.sent` means that the "corresponding" Controller buffer is "destroyed". That can be by Num Complete, Disconnect Complete, HW error or HCI reset. It does not mean the message has been sent anywhere.

This does mean `bt_l2cap_chan_ops.disconnected` could be delayed by the Controller withholding Num Complete, even if the Host has already received an L2CAP disconnect request.

**Comment by Thalley** [comment]
@alwa-nordic I'm not an expert in L2CAP CoC, so I need some things clarified first.

1) Is it valid to terminate a CoC while there is an outstanding read or write? 
2) If it is valid, does the host wait for the credits from remote or controller, or does it reset the credits for the next connection? 

Ideally we would treat CoC similar to ACL and ISO where once they have been terminated, we don't expect any future events for the connections and can reset them once the disconnect event has been received, which allows us to clear the pending TX/RX and return callbacks. This is also how we treat the static L2CAP channels IIRC.

**Comment by alwa-nordic** [comment]
> Is it valid to terminate a CoC while there is an outstanding read or write?

Let's reach for a stateless description without the "when". Since ACL packets (in the same direction) are completely ordered on a connection, we could say:

A `L2CAP_DISCONNECTION_REQ` or `L2CAP_DISCONNECTION_RSP` in a ACL stream invalidates the channel's handle at that point in that stream. Not a point in time, but a point in the stream of packets. This happens independently in the RX and TX directions. The handle is still valid in packets that are ordered before, but it's not possible to refer to that channel in later packets, so in effect, no data can be sent after and no signaling packets can refer to it either.

We can legally send an L2CAP disconnect at any time, and it is, in a clear sense, ordered after any data we already sent.

Higher layer protocols may have some concept of "reads" or "writes", and impose further restrictions. But at L2CAP level, there are no data transactions.

It may also interest you that the Core spec is says the following on dropping data on L2CAP disconnection:

> Once an L2CAP_DISCONNECTION_REQ packet is issued, all incoming data in transit on this L2CAP channel shall be discarded and any new additional outgoing data shall be discarded. Once an L2CAP_DISCONNECTION_REQ packet for a channel has been received, all data queued to be sent out on that channel shall be discarded.

[Core 3.A.4.6 v6.0](https://www.bluetooth.com/wp-content/uploads/Files/Specification/HTML/Core-60/out/en/host/logical-link-control-and-adaptation-protocol-specification.html#UUID-840cebec-bd61-45ce-696d-213fc4feebbc:~:text=Once%20an%20L2CAP_DISCONNECTION_REQ%20packet%20is%20issued%2C%20all%20incoming%20data%20in%20transit%20on%20this%20L2CAP%20channel%20shall%20be%20discarded%20and%20any%20new%20additional%20outgoing%20data%20shall%20be%20discarded.%20Once%20an%20L2CAP_DISCONNECTION_REQ%20packet%20for%20a%20channel%20has%20been%20received%2C%20all%20data%20queued%20to%20be%20sent%20out%20on%20that%20channel%20shall%20be%20discarded.)

Assuming "issued" refers to sequencing the TX, it allows the Host to start dropping incoming packets immediately when it wants to close the channel. This is good for error recovery. OTOH, the as-if rule allows the Host to choose to receive and give any remaining RX packets to the application, so long as it can pretend it happened before the `L2CAP_DISCONNECTION_REQ` in the eyes of the remote. It's not possible to send more credits though, since the handle is invalidated in the TX stream.

Likewise for the reception, it gives a lot of freedom to the implementation on how much to drop. But again, since remote can never refer to this channel again, once credits run out, the rest of the data must be dropped. Oh, and it would be against the as-if rule to send something on the channel as an effect of receiving the disconnect, like "thanks for the disconnect".

> If it is valid, does the host wait for the credits from remote or controller, or does it reset the credits for the next connection?

L2CAP credits scoped to a channel and lost on L2CAP disconnect. We can think of it as the channel and credits still existing in the stateless description, but provably unreachable and irrelevant.

> Ideally we would treat CoC similar to ACL and ISO where once they have been terminated, we don't expect any future events for the connections and can reset them once the disconnect event has been received, which allows us to clear the pending TX/RX and return callbacks. This is also how we treat the static L2CAP channels IIRC.

I think I agree. It seems natural that disconnection is the last event. OTOH, if there is a further "destroyed" event, then that can be though of as the true disconnect. Like TCP has `shutdown()` vs `close()`. I am interested in this discussion. It looks like our L2CAP API does have `bt_l2cap_chan.destroy`, just not documented. I'm not even sure if it's public API.

**Comment by Thalley** [comment]
> all data queued to be sent out on that channel shall be discarded.

The BSIM test is failing to how this part changed. Previously when the data was discarded, it triggered the `sent` callback, which is arguably semantically incorrect but is nice for flow control in the upper layers. 

> It looks like our L2CAP API does have bt_l2cap_chan.destroy, just not documented. I'm not even sure it's public API.

I don't think that's for upper layers to use, given the lack of documentation. We do however have the 
```c
	/* @brief Channel released callback
	 *
	 * If this callback is set it is called when the stack has release all
	 * references to the channel object.
	 */
	void (*released)(struct bt_l2cap_chan *chan);
```

Callback, which may be similar, but not too sure. 

I think what OTS wants/expects is that when a CoC is no longer connected, it does this check to ensure that we are in a similar state as we were before the CoC was connected. Maybe it should be using the `released`, assuming that `released` is call at some point after the disconnection. Not sure how/when it is called.

**Comment by alwa-nordic** [comment]
> ```c
> 	/* @brief Channel released callback
> 	 *
> 	 * If this callback is set it is called when the stack has release all
> 	 * references to the channel object.
> 	 */
> 	void (*released)(struct bt_l2cap_chan *chan);
> ```

Thanks! I read that documentation as saying the invocation of `released` is the last time the Host will reference `chan`. Once `released` is entered, the application can free `chan` or reuse it for the next L2CAP connection. The application cannot free `chan` before. It implies `disconnected` has already been invoked, or that it never will be. And it can only come after all `sent` callbacks.

If the `disconnected` or `sent` callbacks come after `released`, it would be a clear API violation.

> I think what OTS wants/expects is that when a CoC is no longer connected, it does this check to ensure that we are in a similar state as we were before the CoC was connected. Maybe it should be using the `released`, assuming that `released` is call at some point after the disconnection. Not sure how/when it is called.

Yes, to reach the same state as before `bt_l2cap_chan_connect` or `bt_l2cap_server.accept` we must wait for `released`. In between those points, the Host requires `chan` to be alive.

**Comment by alwa-nordic** [comment]
> > all data queued to be sent out on that channel shall be discarded.
> 
> The BSIM test is failing to how this part changed. Previously when the data was discarded, it triggered the `sent` callback, which is arguably semantically incorrect but is nice for flow control in the upper layers.

Revised a bit:

  - `bt_l2cap_chan_ops.sent` must be invoked exactly once for each successful return from `bt_l2cap_chan_send`.
  - `bt_l2cap_chan_ops.sent` must not come after `bt_l2cap_chan_ops.disconnected`.
  - bt_l2cap_chan_ops.sent means either that the "corresponding" Controller buffer is "destroyed". That can be by Num Complete, Disconnect Complete, HW error or HCI reset. Or it means the channel is disconnected. It does not mean the message has been sent anywhere. It does not mean the Controller has freed a buffer, unless the channel state is connected.

The meaning of `bt_l2cap_chan_ops.sent` got a bit watered down. It this ok?

**Comment by Thalley** [comment]
> The meaning of `bt_l2cap_chan_ops.sent` got a bit watered down. It this ok?

I think what you are describing is the existing behavior before this PR, right? 

One issue that the `sent` callback has when compared to e.g. `bt_gatt_write_func_t`, is that it does not contain a `status` or `err` argument. It is not possible for the application to determine if the send was finished successfully, or aborted (which I guess is pretty similar to num_complete_pkts). 

The `sent` callback is defined as 
```c
	/** @brief Channel sent callback
	 *
	 *  This callback will be called once the controller marks the SDU
	 *  as completed. When the controller does so is implementation
	 *  dependent. It could be after the SDU is enqueued for transmission,
	 *  or after it is sent on air.
	 *
	 *  @param chan The channel which has sent data.
	 */
	void (*sent)(struct bt_l2cap_chan *chan);
```

So the current documentation does not define the behavior for what happens at disconnect, and we want to define that we can do it whatever way we think is best. 

> bt_l2cap_chan_ops.sent must be invoked exactly once for each successful return from bt_l2cap_chan_send.

Agree, that would be ideal, but does not meet the current documentation, as it may neither be sent nor enqueue for transmission in the controller, right? Unless I'm forgetting how the host works, there's a chance that it is enqueue in the host, and not yet sent to the controller. 

> bt_l2cap_chan_ops.sent must come before bt_l2cap_chan_ops.disconnected.

This would also be ideal, and make the L2CAP API more similar to ISO and ATT/GATT.

**Comment by alwa-nordic** [comment]
> > The meaning of `bt_l2cap_chan_ops.sent` got a bit watered down. It this ok?
> 
> I think what you are describing is the existing behavior before this PR, right?

Maybe. I was thinking we should first decide on a behavior we want, that makes sense. Once we have what we want we can look at how it's different from what we have.

> One issue that the `sent` callback has when compared to e.g. `bt_gatt_write_func_t`, is that it does not contain a `status` or `err` argument. It is not possible for the application to determine if the send was finished successfully, or aborted (which I guess is pretty similar to num_complete_pkts).

I agree this is not optimal. Actually, `bt_gatt_write_func_t` is a bad example. It also has no way to know that the ATT write did not happen. `err` is the ATT response code. But what if there is no response? Seems we have a tendency to forget error handling in Zephyr. :P

> The `sent` callback is defined as
> 
> ```c
> 	/** @brief Channel sent callback
> 	 *
> 	 *  This callback will be called once the controller marks the SDU
> 	 *  as completed. When the controller does so is implementation
> 	 *  dependent. It could be after the SDU is enqueued for transmission,
> 	 *  or after it is sent on air.
> 	 *
> 	 *  @param chan The channel which has sent data.
> 	 */
> 	void (*sent)(struct bt_l2cap_chan *chan);
> ```
> 
> So the current documentation does not define the behavior for what happens at disconnect, and we want to define that we can do it whatever way we think is best.
> 
> > bt_l2cap_chan_ops.sent must be invoked exactly once for each successful return from bt_l2cap_chan_send.
> 
> Agree, that would be ideal, but does not meet the current documentation, as it may neither be sent nor enqueue for transmission in the controller, right? Unless I'm forgetting how the host works, there's a chance that it is enqueue in the host, and not yet sent to the controller.

Correct. We would have to change the definition of `sent`.

> > bt_l2cap_chan_ops.sent must come before bt_l2cap_chan_ops.disconnected.
> 
> This would also be ideal, and make the L2CAP API more similar to ISO and ATT/GATT.

I can make a PR to ensure this. It's going to require some work to do it cleanly though.

Apart from that, there are some other details to decide on:

  - When does `bt_l2cap_le_chan.state` change on during disconnect? It has states `BT_L2CAP_DISCONNECTING` and `BT_L2CAP_DISCONNECTED`. Should we say it enters `BT_L2CAP_DISCONNECTING` at the point Host before the Host starts discarding data, and `BT_L2CAP_DISCONNECTED` when we are ready to invoke `bt_l2cap_chan_ops.disconnected`?
  - `bt_l2cap_chan_ops.status` has a `BT_L2CAP_STATUS_SHUTDOWN` bit. Changes to this field are notified by `bt_l2cap_chan_ops.status`. Is this relevant to disconnect, and if so, when is it set?

**Comment by alwa-nordic** [comment]
> Previously when the data was discarded, it triggered the `sent` callback, which is arguably semantically incorrect but is nice for flow control in the upper layers.

Does OTS use `sent` for flow control? I think `sent` is not good for flow control. We should use credits instead.

**Comment by Thalley** [comment]
> > Previously when the data was discarded, it triggered the `sent` callback, which is arguably semantically incorrect but is nice for flow control in the upper layers.
> 
> Does OTS use `sent` for flow control? I think `sent` is not good for flow control. We should use credits instead.

I'm not familiar with the inner workings for L2CAP, but since there isn't a `credit` callback when sending, isn't `sent` the callback that indicates a returned credit? 

Would be great to get input from @jhedberg in this PR as well. I'm too unfamiliar with L2CAP to properly determine how it should be used and how it is implemented at this point. 

If @ekohandel is still around, they can perhaps also share their initial thoughts behind OTS's use of L2CAP

**Comment by alwa-nordic** [comment]
> I'm not familiar with the inner workings for L2CAP, but since there isn't a `credit` callback when sending, isn't `sent` the callback that indicates a returned credit?

The `sent` callbacks does not say anything about availability of credits. The current number of credits is stored in `bt_l2cap_le_chan.tx.credits`. I think we should either add a `credit` callback (or we could bake it into `bt_l2cap_chan_ops.status`).

BTW, I think it's best not to think of credits as "returned". They are completely one-time use. They are given by the data receiver and used up by the data sender.

**Comment by ekohandel** [comment]
> If @ekohandel is still around, they can perhaps also share their initial thoughts behind OTS's use of L2CAP

Sorry for the very late reply on this. It has been a while and I had to look through the code again to remind myself of how OTS uses L2CAP.

In the read procedure case the OTS layer will ask the application layer for the data but it may not get all data that it is asking for so it will send as much as it gets from the application. It uses the `sent` callback to either ask application for more data or determine that all data was sent and the procedure is done. I guess this could be considered flow control but it is flow control at application layer and not at the L2CAP layer.

Looking at this more closely there seems to be a flaw in the `oacp_read_proc_cb` function where there is no handling for the L2CAP connection close. The `ocap_write_proc_cb` function sets the `bt_gatt_ots_l2cap.closed` callback and uses that to reset the `cur_obj->state.type` to `BT_GATT_OTS_OBJECT_IDLE_STATE`. Adding that to `oacp_read_proc_cb` may address the immediate failure.

This assumes that the `closed` callback is always called before `disconnected` is called. Is that a safe assumption?

As an aside, I think there might be yet another flaw in OTS. The current implementation assumes that an object read or write is either fully completed or the connection is disconnected during that read or write procedure to reset the object state to `BT_GATT_OTS_OBJECT_IDLE_STATE`. The following scenario could pose an interesting behavior:
1. Client starts an OTS write procedure on an object but doesn't send all expected data through the L2CAP connection so object stays locked
2. Client switches to a different object
3. Client switches back to the original object and wants to start another write but the object is still locked and a new write cannot be started.

I would need to spend some time writing a test to reproduce this scenario as it would be hard to do that in normal conditions.

**Comment by alwa-nordic** [comment]
I want to us to improve our L2CAP interface. I'm looking for ways to put it on a better foundation. RFC.

> It uses the sent callback to either ask application for more data or determine that all data was sent and the procedure is done.

The `sent` callback is poorly defined without relying on Controller specific behavior. I think we should avoid `sent` until someone gives a satisfactory definition that we document, and then use `sent` only on the basis of that definition.

I think it would be better if OTS asks for more data as a reaction to receiving credits or when buffers are free, as that is that that determines how much data can be sent.

To avoid `sent`, could we instead consider the procedure done at a different point? Maybe there is something in the OTS protocol that tells us when the procedure is complete?

If the use of `sent` is meant as flow control, then I think a service alone like OTS is not in a good position to understand the needs of the whole system and manage its own flow control.

The Host limits the use of Controller per ACL connection, so the application does not need to worry about that.

Within a single connection the we can consider adding QoS features for L2CAP TX too, if it's has use cases. But remote can also manage fairness between L2CAP channels using credits and higher layer flow control.

**Review by Thalley — COMMENTED** [review]
@alwa-nordic try to rebase this PR: I think the OTS issue was fixed by https://github.com/zephyrproject-rtos/zephyr/pull/81309

**Comment by jhedberg** [comment]
@alwa-nordic what's the story with this one?

**Comment by alwa-nordic** [comment]
To proceed with this we need to establish whether our API shall invoke all `bt_l2cap_chan_ops.sent` before `bt_l2cap_chan_ops.disconnected` for a given channel. What's you take on this @jhedberg?

**Comment by jhedberg** [comment]
> To proceed with this we need to establish whether our API shall invoke all `bt_l2cap_chan_ops.sent` before `bt_l2cap_chan_ops.disconnected` for a given channel. What's you take on this @jhedberg?

That sounds like the most intuitive behavior, following "the rule of least surprise". You can't send anything over a disconnected channel, so we shouldn't have our API appear to be indicating that.

**Comment by alwa-nordic** [comment]
> https://github.com/zephyrproject-rtos/zephyr/pull/80606#issuecomment-2784512863

Yeah. I agree we should follow the rule of least surprise. In that case, this PR cannot be merged as-is. I looked a bit into enforcing this ordering in a way that does not create inter-thread dependencies, but it appears to me to be a lot of work. I have put this work off for now.

**Comment by alwa-nordic** [comment]
I will close this PR for now so it's clear we are not going to merge. We can continue the discussion here of course.
